// ix/targets/esnext/cjs/asynciterable/toobservable.js
Object.defineProperty(exports, "__esModule", { value: true });
class BooleanSubscription {
    constructor() {
        this.isUnsubscribed = false;
    }
    unsubscribe() {
        this.isUnsubscribed = true;
    }
}
class AsyncIterableObservable {
    constructor(source) {
        this._source = source;
    }
    subscribe(observer) {
        const subscription = new BooleanSubscription();
        const it = this._source[Symbol.asyncIterator]();
        const f = () => {
            it.next()
                .then(({ value, done }) => {
                if (!subscription.isUnsubscribed) {
                    if (done) {
                        observer.complete();
                    }
                    else {
                        observer.next(value);
                        f();
                    }
                }
            })
                .catch(err => {
                if (!subscription.isUnsubscribed) {
                    observer.error(err);
                }
            });
        };
        f();
        return subscription;
    }
}
function toObservable(source) {
    return new AsyncIterableObservable(source);
}
exports.toObservable = toObservable;
