// ix/targets/esnext/cls/iterable/_refcountlist.js
goog.module('targets.esnext.cls.iterable._refcountlist'); exports = {}; var module = {id: 'targets/esnext/cls/iterable/_refcountlist.js'};/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @ignore
 * @record
 * @template T
 */
function RefCount() { }
exports.RefCount = RefCount;
function RefCount_tsickle_Closure_declarations() {
    /** @type {T} */
    RefCount.prototype.value;
    /** @type {number} */
    RefCount.prototype.count;
}
/**
 * @ignore
 * @record
 * @template T
 */
function IRefCountList() { }
exports.IRefCountList = IRefCountList;
function IRefCountList_tsickle_Closure_declarations() {
    /** @type {function(): void} */
    IRefCountList.prototype.clear;
    /** @type {number} */
    IRefCountList.prototype.count;
    /** @type {function(number): T} */
    IRefCountList.prototype.get;
    /** @type {function(T): void} */
    IRefCountList.prototype.push;
    /** @type {function(): void} */
    IRefCountList.prototype.done;
}
/**
 * @ignore
 * @template T
 * @implements {IRefCountList}
 */
class MaxRefCountList {
    constructor() {
        this._list = [];
    }
    /**
     * @return {void}
     */
    clear() { this._list = []; }
    /**
     * @return {number}
     */
    get count() { return this._list.length; }
    /**
     * @param {number} index
     * @return {T}
     */
    get(index) { return this._list[index]; }
    /**
     * @param {T} value
     * @return {void}
     */
    push(value) { this._list.push(value); }
    /**
     * @return {void}
     */
    done() { }
}
exports.MaxRefCountList = MaxRefCountList;
function MaxRefCountList_tsickle_Closure_declarations() {
    /** @type {!Array<T>} */
    MaxRefCountList.prototype._list;
}
/**
 * @ignore
 * @template T
 * @implements {IRefCountList}
 */
class RefCountList {
    /**
     * @param {number} readerCount
     */
    constructor(readerCount) {
        this._count = 0;
        this._readerCount = readerCount;
        this._list = new Map();
    }
    /**
     * @return {void}
     */
    clear() { this._list.clear(); }
    /**
     * @return {number}
     */
    get count() { return this._count; }
    /**
     * @return {number}
     */
    get readerCount() { return this._readerCount; }
    /**
     * @param {number} value
     * @return {void}
     */
    set readerCount(value) { this._readerCount = value; }
    /**
     * @return {void}
     */
    done() { this._readerCount--; }
    /**
     * @param {number} index
     * @return {T}
     */
    get(index) {
        if (!this._list.has(index)) {
            throw new Error('Element no longer available in the buffer.');
        }
        const /** @type {!RefCount<T>} */ res = ((this._list.get(index)));
        const /** @type {T} */ val = res.value;
        if (--res.count === 0) {
            this._list.delete(index);
        }
        return val;
    }
    /**
     * @param {T} value
     * @return {void}
     */
    push(value) {
        this._list.set(this._count++, { value: value, count: this._readerCount });
    }
}
exports.RefCountList = RefCountList;
function RefCountList_tsickle_Closure_declarations() {
    /** @type {number} */
    RefCountList.prototype._readerCount;
    /** @type {!Map<number, !RefCount<T>>} */
    RefCountList.prototype._list;
    /** @type {number} */
    RefCountList.prototype._count;
}
