// ix/targets/es5/cjs/asynciterable/toobservable.js
Object.defineProperty(exports, "__esModule", { value: true });
var BooleanSubscription = (function () {
    function BooleanSubscription() {
        this.isUnsubscribed = false;
    }
    BooleanSubscription.prototype.unsubscribe = function () {
        this.isUnsubscribed = true;
    };
    return BooleanSubscription;
}());
var AsyncIterableObservable = (function () {
    function AsyncIterableObservable(source) {
        this._source = source;
    }
    AsyncIterableObservable.prototype.subscribe = function (observer) {
        var subscription = new BooleanSubscription();
        var it = this._source[Symbol.asyncIterator]();
        var f = function () {
            it.next()
                .then(function (_a) {
                var value = _a.value, done = _a.done;
                if (!subscription.isUnsubscribed) {
                    if (done) {
                        observer.complete();
                    }
                    else {
                        observer.next(value);
                        f();
                    }
                }
            })
                .catch(function (err) {
                if (!subscription.isUnsubscribed) {
                    observer.error(err);
                }
            });
        };
        f();
        return subscription;
    };
    return AsyncIterableObservable;
}());
function toObservable(source) {
    return new AsyncIterableObservable(source);
}
exports.toObservable = toObservable;
