// ix/targets/es5/esm/asynciterable/orderby.js
import * as tslib_1 from "tslib";
import { AsyncIterableX } from '../asynciterable';
import { toArray } from './toarray';
import { sorter as defaultSorter } from '../internal/sorter';
var OrderedAsyncIterableBaseX = (function (_super) {
    tslib_1.__extends(OrderedAsyncIterableBaseX, _super);
    function OrderedAsyncIterableBaseX(source) {
        var _this = _super.call(this) || this;
        _this._source = source;
        return _this;
    }
    OrderedAsyncIterableBaseX.prototype[Symbol.asyncIterator] = function () {
        return tslib_1.__asyncGenerator(this, arguments, function _a() {
            var array, len, indices, i, len_1, indices_1, indices_1_1, index, e_1_1, e_1, _a;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, tslib_1.__await(toArray(this._source))];
                    case 1:
                        array = _b.sent();
                        len = array.length;
                        indices = new Array(len);
                        for (i = 0, len_1 = array.length; i < len_1; i++) {
                            indices[i] = i;
                        }
                        indices.sort(this._getSorter(array));
                        _b.label = 2;
                    case 2:
                        _b.trys.push([2, 7, 8, 9]);
                        indices_1 = tslib_1.__values(indices), indices_1_1 = indices_1.next();
                        _b.label = 3;
                    case 3:
                        if (!!indices_1_1.done) return [3 /*break*/, 6];
                        index = indices_1_1.value;
                        return [4 /*yield*/, array[index]];
                    case 4:
                        _b.sent();
                        _b.label = 5;
                    case 5:
                        indices_1_1 = indices_1.next();
                        return [3 /*break*/, 3];
                    case 6: return [3 /*break*/, 9];
                    case 7:
                        e_1_1 = _b.sent();
                        e_1 = { error: e_1_1 };
                        return [3 /*break*/, 9];
                    case 8:
                        try {
                            if (indices_1_1 && !indices_1_1.done && (_a = indices_1.return)) _a.call(indices_1);
                        }
                        finally { if (e_1) throw e_1.error; }
                        return [7 /*endfinally*/];
                    case 9: return [2 /*return*/];
                }
            });
        });
    };
    OrderedAsyncIterableBaseX.prototype.thenBy = function (keySelector, comparer) {
        if (comparer === void 0) { comparer = defaultSorter; }
        /* tslint:disable-next-line: no-use-before-declare */
        return new OrderedAsyncIterableX(this._source, keySelector, comparer, false, this);
    };
    OrderedAsyncIterableBaseX.prototype.thenByDescending = function (keySelector, comparer) {
        if (comparer === void 0) { comparer = defaultSorter; }
        /* tslint:disable-next-line: no-use-before-declare */
        return new OrderedAsyncIterableX(this._source, keySelector, comparer, true, this);
    };
    return OrderedAsyncIterableBaseX;
}(AsyncIterableX));
export { OrderedAsyncIterableBaseX };
var OrderedAsyncIterableX = (function (_super) {
    tslib_1.__extends(OrderedAsyncIterableX, _super);
    function OrderedAsyncIterableX(source, keySelector, comparer, descending, parent) {
        var _this = _super.call(this, source) || this;
        _this._keySelector = keySelector;
        _this._comparer = comparer;
        _this._descending = descending;
        _this._parent = parent;
        return _this;
    }
    OrderedAsyncIterableX.prototype._getSorter = function (elements, next) {
        var keys = elements.map(this._keySelector);
        var comparer = this._comparer;
        var parent = this._parent;
        var descending = this._descending;
        var sorter = function (x, y) {
            var result = comparer(keys[x], keys[y]);
            if (result === 0) {
                return next ? next(x, y) : x - y;
            }
            return descending ? -result : result;
        };
        return parent ? parent._getSorter(elements, sorter) : sorter;
    };
    return OrderedAsyncIterableX;
}(OrderedAsyncIterableBaseX));
export { OrderedAsyncIterableX };
export function orderBy(source, keySelector, comparer) {
    if (comparer === void 0) { comparer = defaultSorter; }
    return new OrderedAsyncIterableX(source, keySelector, comparer, false);
}
export function orderByDescending(source, keySelector, comparer) {
    if (comparer === void 0) { comparer = defaultSorter; }
    return new OrderedAsyncIterableX(source, keySelector, comparer, true);
}
export function thenBy(source, keySelector, comparer) {
    if (comparer === void 0) { comparer = defaultSorter; }
    return new OrderedAsyncIterableX(source._source, keySelector, comparer, false, source);
}
export function thenByDescending(source, keySelector, comparer) {
    if (comparer === void 0) { comparer = defaultSorter; }
    return new OrderedAsyncIterableX(source._source, keySelector, comparer, true, source);
}
