// ix/targets/es5/cls/asynciterable/concat.js
goog.module('targets.es5.cls.asynciterable.concat'); exports = {}; var module = {id: 'targets/es5/cls/asynciterable/concat.js'};/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

var asynciterable_1 = goog.require('targets.es5.cls.asynciterable');
var tsickle_forward_declare_1 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable");
/**
 * @template TSource
 */
var ConcatAllAsyncIterable = (function (_super) {
    __extends(ConcatAllAsyncIterable, _super);
    /**
     * @param {!AsyncIterable<!AsyncIterable<TSource>>} source
     */
    function ConcatAllAsyncIterable(source) {
        var _this = _super.call(this) || this;
        _this._source = source;
        return _this;
    }
    /**
     * @return {!AsyncIterableIterator<TSource>}
     */
    ConcatAllAsyncIterable.prototype[Symbol.asyncIterator] = function () {
        return __asyncGenerator(this, arguments, function _a() {
            var _a, _b, outer, outer_1, outer_1_1, item, e_1_1, e_2_1, e_2, _c, e_1, _d;
            return __generator(this, function (_e) {
                switch (_e.label) {
                    case 0:
                        _e.trys.push([0, 19, 20, 25]);
                        _a = __asyncValues(this._source);
                        _e.label = 1;
                    case 1: return [4 /*yield*/, __await(_a.next())];
                    case 2:
                        if (!(_b = _e.sent(), !_b.done)) return [3 /*break*/, 18];
                        return [4 /*yield*/, __await(_b.value)];
                    case 3:
                        outer = _e.sent();
                        _e.label = 4;
                    case 4:
                        _e.trys.push([4, 11, 12, 17]);
                        outer_1 = __asyncValues(outer);
                        _e.label = 5;
                    case 5: return [4 /*yield*/, __await(outer_1.next())];
                    case 6:
                        if (!(outer_1_1 = _e.sent(), !outer_1_1.done)) return [3 /*break*/, 10];
                        return [4 /*yield*/, __await(outer_1_1.value)];
                    case 7:
                        item = _e.sent();
                        return [4 /*yield*/, item];
                    case 8:
                        _e.sent();
                        _e.label = 9;
                    case 9: return [3 /*break*/, 5];
                    case 10: return [3 /*break*/, 17];
                    case 11:
                        e_1_1 = _e.sent();
                        e_1 = { error: e_1_1 };
                        return [3 /*break*/, 17];
                    case 12:
                        _e.trys.push([12, , 15, 16]);
                        if (!(outer_1_1 && !outer_1_1.done && (_d = outer_1.return))) return [3 /*break*/, 14];
                        return [4 /*yield*/, __await(_d.call(outer_1))];
                    case 13:
                        _e.sent();
                        _e.label = 14;
                    case 14: return [3 /*break*/, 16];
                    case 15:
                        if (e_1) throw e_1.error;
                        return [7 /*endfinally*/];
                    case 16: return [7 /*endfinally*/];
                    case 17: return [3 /*break*/, 1];
                    case 18: return [3 /*break*/, 25];
                    case 19:
                        e_2_1 = _e.sent();
                        e_2 = { error: e_2_1 };
                        return [3 /*break*/, 25];
                    case 20:
                        _e.trys.push([20, , 23, 24]);
                        if (!(_b && !_b.done && (_c = _a.return))) return [3 /*break*/, 22];
                        return [4 /*yield*/, __await(_c.call(_a))];
                    case 21:
                        _e.sent();
                        _e.label = 22;
                    case 22: return [3 /*break*/, 24];
                    case 23:
                        if (e_2) throw e_2.error;
                        return [7 /*endfinally*/];
                    case 24: return [7 /*endfinally*/];
                    case 25: return [2 /*return*/];
                }
            });
        });
    };
    return ConcatAllAsyncIterable;
}(asynciterable_1.AsyncIterableX));
function ConcatAllAsyncIterable_tsickle_Closure_declarations() {
    /** @type {!AsyncIterable<!AsyncIterable<TSource>>} */
    ConcatAllAsyncIterable.prototype._source;
}
/**
 * @template TSource
 */
var ConcatAsyncIterable = (function (_super) {
    __extends(ConcatAsyncIterable, _super);
    /**
     * @param {!Iterable<!AsyncIterable<TSource>>} source
     */
    function ConcatAsyncIterable(source) {
        var _this = _super.call(this) || this;
        _this._source = source;
        return _this;
    }
    /**
     * @return {!AsyncIterableIterator<TSource>}
     */
    ConcatAsyncIterable.prototype[Symbol.asyncIterator] = function () {
        return __asyncGenerator(this, arguments, function _a() {
            var _a, _b, outer, outer_2, outer_2_1, item, e_3_1, e_4_1, e_3, _c, e_4, _d;
            return __generator(this, function (_e) {
                switch (_e.label) {
                    case 0:
                        _e.trys.push([0, 17, 18, 19]);
                        _a = __values(this._source), _b = _a.next();
                        _e.label = 1;
                    case 1:
                        if (!!_b.done) return [3 /*break*/, 16];
                        outer = _b.value;
                        _e.label = 2;
                    case 2:
                        _e.trys.push([2, 9, 10, 15]);
                        outer_2 = __asyncValues(outer);
                        _e.label = 3;
                    case 3: return [4 /*yield*/, __await(outer_2.next())];
                    case 4:
                        if (!(outer_2_1 = _e.sent(), !outer_2_1.done)) return [3 /*break*/, 8];
                        return [4 /*yield*/, __await(outer_2_1.value)];
                    case 5:
                        item = _e.sent();
                        return [4 /*yield*/, item];
                    case 6:
                        _e.sent();
                        _e.label = 7;
                    case 7: return [3 /*break*/, 3];
                    case 8: return [3 /*break*/, 15];
                    case 9:
                        e_3_1 = _e.sent();
                        e_3 = { error: e_3_1 };
                        return [3 /*break*/, 15];
                    case 10:
                        _e.trys.push([10, , 13, 14]);
                        if (!(outer_2_1 && !outer_2_1.done && (_c = outer_2.return))) return [3 /*break*/, 12];
                        return [4 /*yield*/, __await(_c.call(outer_2))];
                    case 11:
                        _e.sent();
                        _e.label = 12;
                    case 12: return [3 /*break*/, 14];
                    case 13:
                        if (e_3) throw e_3.error;
                        return [7 /*endfinally*/];
                    case 14: return [7 /*endfinally*/];
                    case 15:
                        _b = _a.next();
                        return [3 /*break*/, 1];
                    case 16: return [3 /*break*/, 19];
                    case 17:
                        e_4_1 = _e.sent();
                        e_4 = { error: e_4_1 };
                        return [3 /*break*/, 19];
                    case 18:
                        try {
                            if (_b && !_b.done && (_d = _a.return)) _d.call(_a);
                        }
                        finally { if (e_4) throw e_4.error; }
                        return [7 /*endfinally*/];
                    case 19: return [2 /*return*/];
                }
            });
        });
    };
    return ConcatAsyncIterable;
}(asynciterable_1.AsyncIterableX));
function ConcatAsyncIterable_tsickle_Closure_declarations() {
    /** @type {!Iterable<!AsyncIterable<TSource>>} */
    ConcatAsyncIterable.prototype._source;
}
/**
 * @template TSource
 * @param {!AsyncIterable<!AsyncIterable<TSource>>} source
 * @return {!tsickle_forward_declare_1.AsyncIterableX<TSource>}
 */
function concatAll(source) {
    return new ConcatAllAsyncIterable(source);
}
exports.concatAll = concatAll;
/**
 * @template TSource
 * @param {!Iterable<!AsyncIterable<TSource>>} source
 * @return {!tsickle_forward_declare_1.AsyncIterableX<TSource>}
 */
function _concatAll(source) {
    return new ConcatAsyncIterable(source);
}
exports._concatAll = _concatAll;
/**
 * @template T
 * @param {!AsyncIterable<T>} source
 * @param {...!AsyncIterable<T>} args
 * @return {!tsickle_forward_declare_1.AsyncIterableX<T>}
 */
function concat(source) {
    var args = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        args[_i - 1] = arguments[_i];
    }
    return new ConcatAsyncIterable(__spread([source], args));
}
exports.concat = concat;
/**
 * @template T
 * @param {...!AsyncIterable<T>} args
 * @return {!tsickle_forward_declare_1.AsyncIterableX<T>}
 */
function concatStatic() {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    return new ConcatAsyncIterable(args);
}
exports.concatStatic = concatStatic;
