// ix/targets/es5/cls/asynciterable/orderby.js
goog.module('targets.es5.cls.asynciterable.orderby'); exports = {}; var module = {id: 'targets/es5/cls/asynciterable/orderby.js'};/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

var asynciterable_1 = goog.require('targets.es5.cls.asynciterable');
var tsickle_forward_declare_1 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable");
var toarray_1 = goog.require('targets.es5.cls.asynciterable.toarray');
var tsickle_forward_declare_2 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.toarray");
var sorter_1 = goog.require('targets.es5.cls.internal.sorter');
var tsickle_forward_declare_3 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.internal.sorter");
/**
 * @abstract
 * @template TSource
 */
var OrderedAsyncIterableBaseX = (function (_super) {
    __extends(OrderedAsyncIterableBaseX, _super);
    /**
     * @param {!AsyncIterable<TSource>} source
     */
    function OrderedAsyncIterableBaseX(source) {
        var _this = _super.call(this) || this;
        _this._source = source;
        return _this;
    }
    /**
     * @return {!AsyncIterableIterator<TSource>}
     */
    OrderedAsyncIterableBaseX.prototype[Symbol.asyncIterator] = function () {
        return __asyncGenerator(this, arguments, function _a() {
            var array, len, indices, i, len_1, indices_1, indices_1_1, index, e_1_1, e_1, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, __await(toarray_1.toArray(this._source))];
                    case 1:
                        array = _b.sent();
                        len = array.length;
                        indices = new Array(len);
                        for (i = 0, len_1 = array.length; i < len_1; i++) {
                            indices[i] = i;
                        }
                        indices.sort(this._getSorter(array));
                        _b.label = 2;
                    case 2:
                        _b.trys.push([2, 7, 8, 9]);
                        indices_1 = __values(indices), indices_1_1 = indices_1.next();
                        _b.label = 3;
                    case 3:
                        if (!!indices_1_1.done) return [3 /*break*/, 6];
                        index = indices_1_1.value;
                        return [4 /*yield*/, array[index]];
                    case 4:
                        _b.sent();
                        _b.label = 5;
                    case 5:
                        indices_1_1 = indices_1.next();
                        return [3 /*break*/, 3];
                    case 6: return [3 /*break*/, 9];
                    case 7:
                        e_1_1 = _b.sent();
                        e_1 = { error: e_1_1 };
                        return [3 /*break*/, 9];
                    case 8:
                        try {
                            if (indices_1_1 && !indices_1_1.done && (_a = indices_1.return)) _a.call(indices_1);
                        }
                        finally { if (e_1) throw e_1.error; }
                        return [7 /*endfinally*/];
                    case 9: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * @template TKey
     * @param {function(TSource): TKey} keySelector
     * @param {function(TKey, TKey): number=} comparer
     * @return {!OrderedAsyncIterableBaseX}
     */
    OrderedAsyncIterableBaseX.prototype.thenBy = function (keySelector, comparer) {
        if (comparer === void 0) { comparer = sorter_1.sorter; }
        /* tslint:disable-next-line: no-use-before-declare */
        return new OrderedAsyncIterableX(this._source, keySelector, comparer, false, this);
    };
    /**
     * @template TKey
     * @param {function(TSource): TKey} keySelector
     * @param {function(TKey, TKey): number=} comparer
     * @return {!OrderedAsyncIterableBaseX}
     */
    OrderedAsyncIterableBaseX.prototype.thenByDescending = function (keySelector, comparer) {
        if (comparer === void 0) { comparer = sorter_1.sorter; }
        /* tslint:disable-next-line: no-use-before-declare */
        return new OrderedAsyncIterableX(this._source, keySelector, comparer, true, this);
    };
    return OrderedAsyncIterableBaseX;
}(asynciterable_1.AsyncIterableX));
exports.OrderedAsyncIterableBaseX = OrderedAsyncIterableBaseX;
function OrderedAsyncIterableBaseX_tsickle_Closure_declarations() {
    /** @type {!AsyncIterable<TSource>} */
    OrderedAsyncIterableBaseX.prototype._source;
    /**
     * @abstract
     * @param {!Array<TSource>} elements
     * @param {function(number, number): number=} next
     * @return {function(number, number): number}
     */
    OrderedAsyncIterableBaseX.prototype._getSorter = function (elements, next) { };
}
/**
 * @template TKey, TSource
 */
var OrderedAsyncIterableX = (function (_super) {
    __extends(OrderedAsyncIterableX, _super);
    /**
     * @param {!AsyncIterable<TSource>} source
     * @param {function(TSource): TKey} keySelector
     * @param {function(TKey, TKey): number} comparer
     * @param {boolean} descending
     * @param {!OrderedAsyncIterableBaseX<TSource>=} parent
     */
    function OrderedAsyncIterableX(source, keySelector, comparer, descending, parent) {
        var _this = _super.call(this, source) || this;
        _this._keySelector = keySelector;
        _this._comparer = comparer;
        _this._descending = descending;
        _this._parent = parent;
        return _this;
    }
    /**
     * @param {!Array<TSource>} elements
     * @param {function(number, number): number=} next
     * @return {function(number, number): number}
     */
    OrderedAsyncIterableX.prototype._getSorter = function (elements, next) {
        var /** @type {!Array<TKey>} */ keys = elements.map(this._keySelector);
        var /** @type {function(TKey, TKey): number} */ comparer = this._comparer;
        var /** @type {!OrderedAsyncIterableBaseX<TSource>} */ parent = this._parent;
        var /** @type {boolean} */ descending = this._descending;
        var /** @type {function(number, number): number} */ sorter = function (x, y) {
            var /** @type {number} */ result = comparer(keys[x], keys[y]);
            if (result === 0) {
                return next ? next(x, y) : x - y;
            }
            return descending ? -result : result;
        };
        return parent ? parent._getSorter(elements, sorter) : sorter;
    };
    return OrderedAsyncIterableX;
}(OrderedAsyncIterableBaseX));
exports.OrderedAsyncIterableX = OrderedAsyncIterableX;
function OrderedAsyncIterableX_tsickle_Closure_declarations() {
    /** @type {function(TSource): TKey} */
    OrderedAsyncIterableX.prototype._keySelector;
    /** @type {function(TKey, TKey): number} */
    OrderedAsyncIterableX.prototype._comparer;
    /** @type {boolean} */
    OrderedAsyncIterableX.prototype._descending;
    /** @type {!OrderedAsyncIterableBaseX<TSource>} */
    OrderedAsyncIterableX.prototype._parent;
}
/**
 * @template TKey, TSource
 * @param {!AsyncIterable<TSource>} source
 * @param {function(TSource): TKey} keySelector
 * @param {function(TKey, TKey): number=} comparer
 * @return {!OrderedAsyncIterableX<TKey, TSource>}
 */
function orderBy(source, keySelector, comparer) {
    if (comparer === void 0) { comparer = sorter_1.sorter; }
    return new OrderedAsyncIterableX(source, keySelector, comparer, false);
}
exports.orderBy = orderBy;
/**
 * @template TKey, TSource
 * @param {!AsyncIterable<TSource>} source
 * @param {function(TSource): TKey} keySelector
 * @param {function(TKey, TKey): number=} comparer
 * @return {!OrderedAsyncIterableX<TKey, TSource>}
 */
function orderByDescending(source, keySelector, comparer) {
    if (comparer === void 0) { comparer = sorter_1.sorter; }
    return new OrderedAsyncIterableX(source, keySelector, comparer, true);
}
exports.orderByDescending = orderByDescending;
/**
 * @template TKey, TSource
 * @param {!OrderedAsyncIterableBaseX<TSource>} source
 * @param {function(TSource): TKey} keySelector
 * @param {function(TKey, TKey): number=} comparer
 * @return {!OrderedAsyncIterableX<TKey, TSource>}
 */
function thenBy(source, keySelector, comparer) {
    if (comparer === void 0) { comparer = sorter_1.sorter; }
    return new OrderedAsyncIterableX(source._source, keySelector, comparer, false, source);
}
exports.thenBy = thenBy;
/**
 * @template TKey, TSource
 * @param {!OrderedAsyncIterableBaseX<TSource>} source
 * @param {function(TSource): TKey} keySelector
 * @param {function(TKey, TKey): number=} comparer
 * @return {!OrderedAsyncIterableX<TKey, TSource>}
 */
function thenByDescending(source, keySelector, comparer) {
    if (comparer === void 0) { comparer = sorter_1.sorter; }
    return new OrderedAsyncIterableX(source._source, keySelector, comparer, true, source);
}
exports.thenByDescending = thenByDescending;
