// ix/targets/es5/cls/asynciterable/timeout.js
goog.module('targets.es5.cls.asynciterable.timeout'); exports = {}; var module = {id: 'targets/es5/cls/asynciterable/timeout.js'};/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

var asynciterable_1 = goog.require('targets.es5.cls.asynciterable');
var tsickle_forward_declare_1 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable");
var _sleep_1 = goog.require('targets.es5.cls.asynciterable._sleep');
var tsickle_forward_declare_2 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable._sleep");
var TimeoutError = (function (_super) {
    __extends(TimeoutError, _super);
    function TimeoutError() {
        var _this = _super.call(this) || this;
        Object.setPrototypeOf(_this, TimeoutError.prototype);
        _this.message = 'Timeout has occurred';
        return _this;
    }
    return TimeoutError;
}(Error));
exports.TimeoutError = TimeoutError;
var /** @type {string} */ VALUE_TYPE = 'value';
var /** @type {string} */ ERROR_TYPE = 'error';
/**
 * @record
 * @template T
 */
function TimeoutOperation() { }
function TimeoutOperation_tsickle_Closure_declarations() {
    /** @type {string} */
    TimeoutOperation.prototype.type;
    /** @type {!IteratorResult<T>} */
    TimeoutOperation.prototype.value;
}
/**
 * @template TSource
 */
var TimeoutAsyncIterable = (function (_super) {
    __extends(TimeoutAsyncIterable, _super);
    /**
     * @param {!AsyncIterable<TSource>} source
     * @param {number} dueTime
     */
    function TimeoutAsyncIterable(source, dueTime) {
        var _this = _super.call(this) || this;
        _this._source = source;
        _this._dueTime = dueTime;
        return _this;
    }
    /**
     * @return {!AsyncIterableIterator<TSource>}
     */
    TimeoutAsyncIterable.prototype[Symbol.asyncIterator] = function () {
        return __asyncGenerator(this, arguments, function _a() {
            var it, _a, type, value;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        it = this._source[Symbol.asyncIterator]();
                        _b.label = 1;
                    case 1:
                        if (!1) return [3 /*break*/, 4];
                        return [4 /*yield*/, __await(Promise.race([
                                it.next().then(function (value) { return { type: VALUE_TYPE, value: value }; }),
                                _sleep_1.sleep(this._dueTime).then(function () { return { type: ERROR_TYPE }; })
                            ]))];
                    case 2:
                        _a = _b.sent(), type = _a.type, value = _a.value;
                        if (type === ERROR_TYPE) {
                            throw new TimeoutError();
                        }
                        if (value.done) {
                            return [3 /*break*/, 4];
                        }
                        return [4 /*yield*/, value.value];
                    case 3:
                        _b.sent();
                        return [3 /*break*/, 1];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    return TimeoutAsyncIterable;
}(asynciterable_1.AsyncIterableX));
function TimeoutAsyncIterable_tsickle_Closure_declarations() {
    /** @type {!AsyncIterable<TSource>} */
    TimeoutAsyncIterable.prototype._source;
    /** @type {number} */
    TimeoutAsyncIterable.prototype._dueTime;
}
/**
 * @template TSource
 * @param {!AsyncIterable<TSource>} source
 * @param {number} dueTime
 * @return {!tsickle_forward_declare_1.AsyncIterableX<TSource>}
 */
function timeout(source, dueTime) {
    return new TimeoutAsyncIterable(source, dueTime);
}
exports.timeout = timeout;
