// ix/targets/es5/cls/asynciterable/toobservable.js
goog.module('targets.es5.cls.asynciterable.toobservable'); exports = {}; var module = {id: 'targets/es5/cls/asynciterable/toobservable.js'};/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

var tsickle_forward_declare_1 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.observer");
goog.require("_Users.ptaylor.dev.ixjs.src.observer"); // force type-only module to be loaded
var tsickle_forward_declare_2 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.subscription");
goog.require("_Users.ptaylor.dev.ixjs.src.subscription"); // force type-only module to be loaded
/**
 * @implements {tsickle_forward_declare_2.Subscription}
 */
var BooleanSubscription = (function () {
    function BooleanSubscription() {
        this.isUnsubscribed = false;
    }
    /**
     * @return {void}
     */
    BooleanSubscription.prototype.unsubscribe = function () {
        this.isUnsubscribed = true;
    };
    return BooleanSubscription;
}());
function BooleanSubscription_tsickle_Closure_declarations() {
    /** @type {boolean} */
    BooleanSubscription.prototype.isUnsubscribed;
}
/**
 * @template TSource
 * @implements {tsickle_forward_declare_1.Observable}
 */
var AsyncIterableObservable = (function () {
    /**
     * @param {!AsyncIterable<TSource>} source
     */
    function AsyncIterableObservable(source) {
        this._source = source;
    }
    /**
     * @param {!tsickle_forward_declare_1.Observer<TSource>} observer
     * @return {!BooleanSubscription}
     */
    AsyncIterableObservable.prototype.subscribe = function (observer) {
        var /** @type {!BooleanSubscription} */ subscription = new BooleanSubscription();
        var /** @type {!AsyncIterator<TSource>} */ it = this._source[Symbol.asyncIterator]();
        var /** @type {function(): void} */ f = function () {
            it.next()
                .then(function (_a) {
                var value = _a.value, done = _a.done;
                if (!subscription.isUnsubscribed) {
                    if (done) {
                        observer.complete();
                    }
                    else {
                        observer.next(value);
                        f();
                    }
                }
            })
                .catch(function (err) {
                if (!subscription.isUnsubscribed) {
                    observer.error(err);
                }
            });
        };
        f();
        return subscription;
    };
    return AsyncIterableObservable;
}());
function AsyncIterableObservable_tsickle_Closure_declarations() {
    /** @type {!AsyncIterable<TSource>} */
    AsyncIterableObservable.prototype._source;
}
/**
 * @template TSource
 * @param {!AsyncIterable<TSource>} source
 * @return {!tsickle_forward_declare_1.Observable<TSource>}
 */
function toObservable(source) {
    return new AsyncIterableObservable(source);
}
exports.toObservable = toObservable;
