// ix/targets/es5/cjs/scheduler.js
Object.defineProperty(exports, "__esModule", { value: true });
var ActionSubscription = (function () {
    function ActionSubscription(action) {
        this.isUnsubscribed = false;
        this._action = action;
    }
    ActionSubscription.prototype.unsubscribe = function () {
        if (!this.isUnsubscribed) {
            this.isUnsubscribed = true;
            this._action();
        }
    };
    return ActionSubscription;
}());
var DefaultScheduler = (function () {
    function DefaultScheduler() {
    }
    Object.defineProperty(DefaultScheduler.prototype, "now", {
        get: function () {
            return Date.now();
        },
        enumerable: true,
        configurable: true
    });
    DefaultScheduler.prototype.delay = function (dueTime) {
        return new Promise(function (res) { return setTimeout(res, dueTime); });
    };
    DefaultScheduler.prototype.schedule = function (action, dueTime) {
        var id = setTimeout(function () { return action(); }, dueTime);
        return new ActionSubscription(function () { return clearTimeout(id); });
    };
    return DefaultScheduler;
}());
exports.DefaultScheduler = DefaultScheduler;
