// ix/targets/esnext/cls/asynciterable/timeout.js
goog.module('targets.esnext.cls.asynciterable.timeout'); exports = {}; var module = {id: 'targets/esnext/cls/asynciterable/timeout.js'};/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

var asynciterable_1 = goog.require('targets.esnext.cls.asynciterable');
const tsickle_forward_declare_1 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable");
var _sleep_1 = goog.require('targets.esnext.cls.asynciterable._sleep');
const tsickle_forward_declare_2 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable._sleep");
class TimeoutError extends Error {
    constructor() {
        super();
        Object.setPrototypeOf(this, TimeoutError.prototype);
        this.message = 'Timeout has occurred';
    }
}
exports.TimeoutError = TimeoutError;
const /** @type {string} */ VALUE_TYPE = 'value';
const /** @type {string} */ ERROR_TYPE = 'error';
/**
 * @record
 * @template T
 */
function TimeoutOperation() { }
function TimeoutOperation_tsickle_Closure_declarations() {
    /** @type {string} */
    TimeoutOperation.prototype.type;
    /** @type {!IteratorResult<T>} */
    TimeoutOperation.prototype.value;
}
/**
 * @template TSource
 */
class TimeoutAsyncIterable extends asynciterable_1.AsyncIterableX {
    /**
     * @param {!AsyncIterable<TSource>} source
     * @param {number} dueTime
     */
    constructor(source, dueTime) {
        super();
        this._source = source;
        this._dueTime = dueTime;
    }
    /**
     * @return {!AsyncIterableIterator<TSource>}
     */
    async *[Symbol.asyncIterator]() {
        const /** @type {!AsyncIterator<TSource>} */ it = this._source[Symbol.asyncIterator]();
        while (1) {
            const { type, value } = await Promise.race([
                it.next().then(value => { return { type: VALUE_TYPE, value }; }),
                _sleep_1.sleep(this._dueTime).then(() => { return { type: ERROR_TYPE }; })
            ]);
            if (type === ERROR_TYPE) {
                throw new TimeoutError();
            }
            if (value.done) {
                break;
            }
            yield value.value;
        }
    }
}
function TimeoutAsyncIterable_tsickle_Closure_declarations() {
    /** @type {!AsyncIterable<TSource>} */
    TimeoutAsyncIterable.prototype._source;
    /** @type {number} */
    TimeoutAsyncIterable.prototype._dueTime;
}
/**
 * @template TSource
 * @param {!AsyncIterable<TSource>} source
 * @param {number} dueTime
 * @return {!tsickle_forward_declare_1.AsyncIterableX<TSource>}
 */
function timeout(source, dueTime) {
    return new TimeoutAsyncIterable(source, dueTime);
}
exports.timeout = timeout;
