// ix/targets/es5/cls/scheduler.js
goog.module('targets.es5.cls.scheduler'); exports = {}; var module = {id: 'targets/es5/cls/scheduler.js'};/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

var tsickle_forward_declare_1 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.subscription");
goog.require("_Users.ptaylor.dev.ixjs.src.subscription"); // force type-only module to be loaded
/**
 * @record
 */
function Scheduler() { }
exports.Scheduler = Scheduler;
function Scheduler_tsickle_Closure_declarations() {
    /** @type {number} */
    Scheduler.prototype.now;
    /** @type {function(number): !Promise<void>} */
    Scheduler.prototype.delay;
    /** @type {function(function(): void, number): !tsickle_forward_declare_1.Subscription} */
    Scheduler.prototype.schedule;
}
/**
 * @implements {tsickle_forward_declare_1.Subscription}
 */
var ActionSubscription = (function () {
    /**
     * @param {function(): void} action
     */
    function ActionSubscription(action) {
        this.isUnsubscribed = false;
        this._action = action;
    }
    /**
     * @return {void}
     */
    ActionSubscription.prototype.unsubscribe = function () {
        if (!this.isUnsubscribed) {
            this.isUnsubscribed = true;
            this._action();
        }
    };
    return ActionSubscription;
}());
function ActionSubscription_tsickle_Closure_declarations() {
    /** @type {boolean} */
    ActionSubscription.prototype.isUnsubscribed;
    /** @type {function(): void} */
    ActionSubscription.prototype._action;
}
/**
 * @implements {Scheduler}
 */
var DefaultScheduler = (function () {
    function DefaultScheduler() {
    }
    Object.defineProperty(DefaultScheduler.prototype, "now", {
        /**
         * @return {number}
         */
        get: function () {
            return Date.now();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {number} dueTime
     * @return {!Promise<void>}
     */
    DefaultScheduler.prototype.delay = function (dueTime) {
        return new Promise(function (res) { return setTimeout(res, dueTime); });
    };
    /**
     * @param {function(): void} action
     * @param {number} dueTime
     * @return {!tsickle_forward_declare_1.Subscription}
     */
    DefaultScheduler.prototype.schedule = function (action, dueTime) {
        var /** @type {!NodeJS.Timer} */ id = setTimeout(function () { return action(); }, dueTime);
        return new ActionSubscription(function () { return clearTimeout(id); });
    };
    return DefaultScheduler;
}());
exports.DefaultScheduler = DefaultScheduler;
