// ix/targets/es2015/cls/scheduler.js
goog.module('targets.es2015.cls.scheduler'); exports = {}; var module = {id: 'targets/es2015/cls/scheduler.js'};/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

const tsickle_forward_declare_1 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.subscription");
goog.require("_Users.ptaylor.dev.ixjs.src.subscription"); // force type-only module to be loaded
/**
 * @record
 */
function Scheduler() { }
exports.Scheduler = Scheduler;
function Scheduler_tsickle_Closure_declarations() {
    /** @type {number} */
    Scheduler.prototype.now;
    /** @type {function(number): !Promise<void>} */
    Scheduler.prototype.delay;
    /** @type {function(function(): void, number): !tsickle_forward_declare_1.Subscription} */
    Scheduler.prototype.schedule;
}
/**
 * @implements {tsickle_forward_declare_1.Subscription}
 */
class ActionSubscription {
    /**
     * @param {function(): void} action
     */
    constructor(action) {
        this.isUnsubscribed = false;
        this._action = action;
    }
    /**
     * @return {void}
     */
    unsubscribe() {
        if (!this.isUnsubscribed) {
            this.isUnsubscribed = true;
            this._action();
        }
    }
}
function ActionSubscription_tsickle_Closure_declarations() {
    /** @type {boolean} */
    ActionSubscription.prototype.isUnsubscribed;
    /** @type {function(): void} */
    ActionSubscription.prototype._action;
}
/**
 * @implements {Scheduler}
 */
class DefaultScheduler {
    /**
     * @return {number}
     */
    get now() {
        return Date.now();
    }
    /**
     * @param {number} dueTime
     * @return {!Promise<void>}
     */
    delay(dueTime) {
        return new Promise(res => setTimeout(res, dueTime));
    }
    /**
     * @param {function(): void} action
     * @param {number} dueTime
     * @return {!tsickle_forward_declare_1.Subscription}
     */
    schedule(action, dueTime) {
        const /** @type {!NodeJS.Timer} */ id = setTimeout(() => action(), dueTime);
        return new ActionSubscription(() => clearTimeout(id));
    }
}
exports.DefaultScheduler = DefaultScheduler;
