// ix/targets/es5/cls/iterable/orderby.js
goog.module('targets.es5.cls.iterable.orderby'); exports = {}; var module = {id: 'targets/es5/cls/iterable/orderby.js'};/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

var iterable_1 = goog.require('targets.es5.cls.iterable');
var tsickle_forward_declare_1 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.iterable");
var sorter_1 = goog.require('targets.es5.cls.internal.sorter');
var tsickle_forward_declare_2 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.internal.sorter");
/**
 * @abstract
 * @template TSource
 */
var OrderedIterableBaseX = (function (_super) {
    __extends(OrderedIterableBaseX, _super);
    /**
     * @param {!Iterable<TSource>} source
     */
    function OrderedIterableBaseX(source) {
        var _this = _super.call(this) || this;
        _this._source = source;
        return _this;
    }
    /**
     * @return {!IterableIterator<TSource>}
     */
    OrderedIterableBaseX.prototype[Symbol.iterator] = function () {
        var array, len, indices, i, len_1, indices_1, indices_1_1, index, e_1_1, e_1, _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    array = Array.from(this._source);
                    len = array.length;
                    indices = new Array(len);
                    for (i = 0, len_1 = array.length; i < len_1; i++) {
                        indices[i] = i;
                    }
                    indices.sort(this._getSorter(array));
                    _b.label = 1;
                case 1:
                    _b.trys.push([1, 6, 7, 8]);
                    indices_1 = __values(indices), indices_1_1 = indices_1.next();
                    _b.label = 2;
                case 2:
                    if (!!indices_1_1.done) return [3 /*break*/, 5];
                    index = indices_1_1.value;
                    return [4 /*yield*/, array[index]];
                case 3:
                    _b.sent();
                    _b.label = 4;
                case 4:
                    indices_1_1 = indices_1.next();
                    return [3 /*break*/, 2];
                case 5: return [3 /*break*/, 8];
                case 6:
                    e_1_1 = _b.sent();
                    e_1 = { error: e_1_1 };
                    return [3 /*break*/, 8];
                case 7:
                    try {
                        if (indices_1_1 && !indices_1_1.done && (_a = indices_1.return)) _a.call(indices_1);
                    }
                    finally { if (e_1) throw e_1.error; }
                    return [7 /*endfinally*/];
                case 8: return [2 /*return*/];
            }
        });
    };
    /**
     * @template TKey
     * @param {function(TSource): TKey} keySelector
     * @param {function(TKey, TKey): number=} comparer
     * @return {!OrderedIterableBaseX}
     */
    OrderedIterableBaseX.prototype.thenBy = function (keySelector, comparer) {
        if (comparer === void 0) { comparer = sorter_1.sorter; }
        /* tslint:disable-next-line: no-use-before-declare */
        return new OrderedIterableX(this._source, keySelector, comparer, false, this);
    };
    /**
     * @template TKey
     * @param {function(TSource): TKey} keySelector
     * @param {function(TKey, TKey): number=} comparer
     * @return {!OrderedIterableBaseX}
     */
    OrderedIterableBaseX.prototype.thenByDescending = function (keySelector, comparer) {
        if (comparer === void 0) { comparer = sorter_1.sorter; }
        /* tslint:disable-next-line: no-use-before-declare */
        return new OrderedIterableX(this._source, keySelector, comparer, true, this);
    };
    return OrderedIterableBaseX;
}(iterable_1.IterableX));
exports.OrderedIterableBaseX = OrderedIterableBaseX;
function OrderedIterableBaseX_tsickle_Closure_declarations() {
    /** @type {!Iterable<TSource>} */
    OrderedIterableBaseX.prototype._source;
    /**
     * @abstract
     * @param {!Array<TSource>} elements
     * @param {function(number, number): number=} next
     * @return {function(number, number): number}
     */
    OrderedIterableBaseX.prototype._getSorter = function (elements, next) { };
}
/**
 * @template TKey, TSource
 */
var OrderedIterableX = (function (_super) {
    __extends(OrderedIterableX, _super);
    /**
     * @param {!Iterable<TSource>} source
     * @param {function(TSource): TKey} keySelector
     * @param {function(TKey, TKey): number} comparer
     * @param {boolean} descending
     * @param {!OrderedIterableBaseX<TSource>=} parent
     */
    function OrderedIterableX(source, keySelector, comparer, descending, parent) {
        var _this = _super.call(this, source) || this;
        _this._keySelector = keySelector;
        _this._comparer = comparer;
        _this._descending = descending;
        _this._parent = parent;
        return _this;
    }
    /**
     * @param {!Array<TSource>} elements
     * @param {function(number, number): number=} next
     * @return {function(number, number): number}
     */
    OrderedIterableX.prototype._getSorter = function (elements, next) {
        var /** @type {!Array<TKey>} */ keys = elements.map(this._keySelector);
        var /** @type {function(TKey, TKey): number} */ comparer = this._comparer;
        var /** @type {!OrderedIterableBaseX<TSource>} */ parent = this._parent;
        var /** @type {boolean} */ descending = this._descending;
        var /** @type {function(number, number): number} */ sorter = function (x, y) {
            var /** @type {number} */ result = comparer(keys[x], keys[y]);
            if (result === 0) {
                return next ? next(x, y) : x - y;
            }
            return descending ? -result : result;
        };
        return parent ? parent._getSorter(elements, sorter) : sorter;
    };
    return OrderedIterableX;
}(OrderedIterableBaseX));
exports.OrderedIterableX = OrderedIterableX;
function OrderedIterableX_tsickle_Closure_declarations() {
    /** @type {function(TSource): TKey} */
    OrderedIterableX.prototype._keySelector;
    /** @type {function(TKey, TKey): number} */
    OrderedIterableX.prototype._comparer;
    /** @type {boolean} */
    OrderedIterableX.prototype._descending;
    /** @type {!OrderedIterableBaseX<TSource>} */
    OrderedIterableX.prototype._parent;
}
/**
 * @template TKey, TSource
 * @param {!Iterable<TSource>} source
 * @param {function(TSource): TKey} keySelector
 * @param {function(TKey, TKey): number=} comparer
 * @return {!OrderedIterableX<TKey, TSource>}
 */
function orderBy(source, keySelector, comparer) {
    if (comparer === void 0) { comparer = sorter_1.sorter; }
    return new OrderedIterableX(source, keySelector, comparer, false);
}
exports.orderBy = orderBy;
/**
 * @template TKey, TSource
 * @param {!Iterable<TSource>} source
 * @param {function(TSource): TKey} keySelector
 * @param {function(TKey, TKey): number=} comparer
 * @return {!OrderedIterableX<TKey, TSource>}
 */
function orderByDescending(source, keySelector, comparer) {
    if (comparer === void 0) { comparer = sorter_1.sorter; }
    return new OrderedIterableX(source, keySelector, comparer, true);
}
exports.orderByDescending = orderByDescending;
/**
 * @template TKey, TSource
 * @param {!OrderedIterableBaseX<TSource>} source
 * @param {function(TSource): TKey} keySelector
 * @param {function(TKey, TKey): number=} comparer
 * @return {!OrderedIterableX<TKey, TSource>}
 */
function thenBy(source, keySelector, comparer) {
    if (comparer === void 0) { comparer = sorter_1.sorter; }
    return new OrderedIterableX(source._source, keySelector, comparer, false, source);
}
exports.thenBy = thenBy;
/**
 * @template TKey, TSource
 * @param {!OrderedIterableBaseX<TSource>} source
 * @param {function(TSource): TKey} keySelector
 * @param {function(TKey, TKey): number=} comparer
 * @return {!OrderedIterableX<TKey, TSource>}
 */
function thenByDescending(source, keySelector, comparer) {
    if (comparer === void 0) { comparer = sorter_1.sorter; }
    return new OrderedIterableX(source._source, keySelector, comparer, true, source);
}
exports.thenByDescending = thenByDescending;
