// ix/targets/es2015/cls/asynciterable/__modules.js
goog.module('targets.es2015.cls.asynciterable.__modules'); exports = {}; var module = {id: 'targets/es2015/cls/asynciterable/__modules.js'};/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

var average_1 = goog.require('targets.es2015.cls.asynciterable.average');
exports.average = average_1.average;
const tsickle_forward_declare_1 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.average");
var buffer_1 = goog.require('targets.es2015.cls.asynciterable.buffer');
exports.buffer = buffer_1.buffer;
const tsickle_forward_declare_2 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.buffer");
var case_1 = goog.require('targets.es2015.cls.asynciterable.case');
exports._case = case_1._case;
const tsickle_forward_declare_3 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.case");
var catch_1 = goog.require('targets.es2015.cls.asynciterable.catch');
exports._catch = catch_1._catch;
exports._catchStatic = catch_1._catchStatic;
const tsickle_forward_declare_4 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.catch");
var catchwith_1 = goog.require('targets.es2015.cls.asynciterable.catchwith');
exports.catchWith = catchwith_1.catchWith;
const tsickle_forward_declare_5 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.catchwith");
var chain_1 = goog.require('targets.es2015.cls.asynciterable.chain');
exports.chain = chain_1.chain;
const tsickle_forward_declare_6 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.chain");
var concat_1 = goog.require('targets.es2015.cls.asynciterable.concat');
exports.concat = concat_1.concat;
exports.concatAll = concat_1.concatAll;
exports.concatStatic = concat_1.concatStatic;
const tsickle_forward_declare_7 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.concat");
var count_1 = goog.require('targets.es2015.cls.asynciterable.count');
exports.count = count_1.count;
const tsickle_forward_declare_8 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.count");
var create_1 = goog.require('targets.es2015.cls.asynciterable.create');
exports.create = create_1.create;
const tsickle_forward_declare_9 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.create");
var debounce_1 = goog.require('targets.es2015.cls.asynciterable.debounce');
exports.debounce = debounce_1.debounce;
const tsickle_forward_declare_10 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.debounce");
var defaultifempty_1 = goog.require('targets.es2015.cls.asynciterable.defaultifempty');
exports.defaultIfEmpty = defaultifempty_1.defaultIfEmpty;
const tsickle_forward_declare_11 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.defaultifempty");
var defer_1 = goog.require('targets.es2015.cls.asynciterable.defer');
exports.defer = defer_1.defer;
const tsickle_forward_declare_12 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.defer");
var distinct_1 = goog.require('targets.es2015.cls.asynciterable.distinct');
exports.distinct = distinct_1.distinct;
const tsickle_forward_declare_13 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.distinct");
var distinctuntilchanged_1 = goog.require('targets.es2015.cls.asynciterable.distinctuntilchanged');
exports.distinctUntilChanged = distinctuntilchanged_1.distinctUntilChanged;
const tsickle_forward_declare_14 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.distinctuntilchanged");
var dowhile_1 = goog.require('targets.es2015.cls.asynciterable.dowhile');
exports.doWhile = dowhile_1.doWhile;
const tsickle_forward_declare_15 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.dowhile");
var elementat_1 = goog.require('targets.es2015.cls.asynciterable.elementat');
exports.elementAt = elementat_1.elementAt;
const tsickle_forward_declare_16 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.elementat");
var empty_1 = goog.require('targets.es2015.cls.asynciterable.empty');
exports.empty = empty_1.empty;
const tsickle_forward_declare_17 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.empty");
var every_1 = goog.require('targets.es2015.cls.asynciterable.every');
exports.every = every_1.every;
const tsickle_forward_declare_18 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.every");
var except_1 = goog.require('targets.es2015.cls.asynciterable.except');
exports.except = except_1.except;
const tsickle_forward_declare_19 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.except");
var expand_1 = goog.require('targets.es2015.cls.asynciterable.expand');
exports.expand = expand_1.expand;
const tsickle_forward_declare_20 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.expand");
var filter_1 = goog.require('targets.es2015.cls.asynciterable.filter');
exports.filter = filter_1.filter;
const tsickle_forward_declare_21 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.filter");
var finally_1 = goog.require('targets.es2015.cls.asynciterable.finally');
exports._finally = finally_1._finally;
const tsickle_forward_declare_22 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.finally");
var find_1 = goog.require('targets.es2015.cls.asynciterable.find');
exports.find = find_1.find;
const tsickle_forward_declare_23 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.find");
var findindex_1 = goog.require('targets.es2015.cls.asynciterable.findindex');
exports.findIndex = findindex_1.findIndex;
const tsickle_forward_declare_24 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.findindex");
var first_1 = goog.require('targets.es2015.cls.asynciterable.first');
exports.first = first_1.first;
const tsickle_forward_declare_25 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.first");
var flatmap_1 = goog.require('targets.es2015.cls.asynciterable.flatmap');
exports.flatMap = flatmap_1.flatMap;
const tsickle_forward_declare_26 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.flatmap");
var flatten_1 = goog.require('targets.es2015.cls.asynciterable.flatten');
exports.flatten = flatten_1.flatten;
const tsickle_forward_declare_27 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.flatten");
var for_1 = goog.require('targets.es2015.cls.asynciterable.for');
exports._for = for_1._for;
const tsickle_forward_declare_28 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.for");
var foreach_1 = goog.require('targets.es2015.cls.asynciterable.foreach');
exports.forEach = foreach_1.forEach;
const tsickle_forward_declare_29 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.foreach");
var from_1 = goog.require('targets.es2015.cls.asynciterable.from');
exports.from = from_1.from;
const tsickle_forward_declare_30 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.from");
var fromevent_1 = goog.require('targets.es2015.cls.asynciterable.fromevent');
exports.fromEvent = fromevent_1.fromEvent;
const tsickle_forward_declare_31 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.fromevent");
var fromeventpattern_1 = goog.require('targets.es2015.cls.asynciterable.fromeventpattern');
exports.fromEventPattern = fromeventpattern_1.fromEventPattern;
const tsickle_forward_declare_32 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.fromeventpattern");
var generate_1 = goog.require('targets.es2015.cls.asynciterable.generate');
exports.generate = generate_1.generate;
const tsickle_forward_declare_33 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.generate");
var generatetime_1 = goog.require('targets.es2015.cls.asynciterable.generatetime');
exports.generateTime = generatetime_1.generateTime;
const tsickle_forward_declare_34 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.generatetime");
var groupby_1 = goog.require('targets.es2015.cls.asynciterable.groupby');
exports.groupBy = groupby_1.groupBy;
const tsickle_forward_declare_35 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.groupby");
var groupjoin_1 = goog.require('targets.es2015.cls.asynciterable.groupjoin');
exports.groupJoin = groupjoin_1.groupJoin;
const tsickle_forward_declare_36 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.groupjoin");
var if_1 = goog.require('targets.es2015.cls.asynciterable.if');
exports._if = if_1._if;
const tsickle_forward_declare_37 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.if");
var ignoreelements_1 = goog.require('targets.es2015.cls.asynciterable.ignoreelements');
exports.ignoreElements = ignoreelements_1.ignoreElements;
const tsickle_forward_declare_38 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.ignoreelements");
var includes_1 = goog.require('targets.es2015.cls.asynciterable.includes');
exports.includes = includes_1.includes;
const tsickle_forward_declare_39 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.includes");
var innerjoin_1 = goog.require('targets.es2015.cls.asynciterable.innerjoin');
exports.innerJoin = innerjoin_1.innerJoin;
const tsickle_forward_declare_40 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.innerjoin");
var intersect_1 = goog.require('targets.es2015.cls.asynciterable.intersect');
exports.intersect = intersect_1.intersect;
const tsickle_forward_declare_41 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.intersect");
var isempty_1 = goog.require('targets.es2015.cls.asynciterable.isempty');
exports.isEmpty = isempty_1.isEmpty;
const tsickle_forward_declare_42 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.isempty");
var last_1 = goog.require('targets.es2015.cls.asynciterable.last');
exports.last = last_1.last;
const tsickle_forward_declare_43 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.last");
var map_1 = goog.require('targets.es2015.cls.asynciterable.map');
exports.map = map_1.map;
const tsickle_forward_declare_44 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.map");
var max_1 = goog.require('targets.es2015.cls.asynciterable.max');
exports.max = max_1.max;
const tsickle_forward_declare_45 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.max");
var maxby_1 = goog.require('targets.es2015.cls.asynciterable.maxby');
exports.maxBy = maxby_1.maxBy;
const tsickle_forward_declare_46 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.maxby");
var memoize_1 = goog.require('targets.es2015.cls.asynciterable.memoize');
exports.memoize = memoize_1.memoize;
const tsickle_forward_declare_47 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.memoize");
var min_1 = goog.require('targets.es2015.cls.asynciterable.min');
exports.min = min_1.min;
const tsickle_forward_declare_48 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.min");
var minby_1 = goog.require('targets.es2015.cls.asynciterable.minby');
exports.minBy = minby_1.minBy;
const tsickle_forward_declare_49 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.minby");
var of_1 = goog.require('targets.es2015.cls.asynciterable.of');
exports.of = of_1.of;
const tsickle_forward_declare_50 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.of");
var ofentries_1 = goog.require('targets.es2015.cls.asynciterable.ofentries');
exports.ofEntries = ofentries_1.ofEntries;
const tsickle_forward_declare_51 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.ofentries");
var ofkeys_1 = goog.require('targets.es2015.cls.asynciterable.ofkeys');
exports.ofKeys = ofkeys_1.ofKeys;
const tsickle_forward_declare_52 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.ofkeys");
var ofvalues_1 = goog.require('targets.es2015.cls.asynciterable.ofvalues');
exports.ofValues = ofvalues_1.ofValues;
const tsickle_forward_declare_53 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.ofvalues");
var onerrorresumenext_1 = goog.require('targets.es2015.cls.asynciterable.onerrorresumenext');
exports.onErrorResumeNext = onerrorresumenext_1.onErrorResumeNext;
exports.onErrorResumeNextStatic = onerrorresumenext_1.onErrorResumeNextStatic;
const tsickle_forward_declare_54 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.onerrorresumenext");
var orderby_1 = goog.require('targets.es2015.cls.asynciterable.orderby');
exports.orderBy = orderby_1.orderBy;
exports.orderByDescending = orderby_1.orderByDescending;
exports.thenBy = orderby_1.thenBy;
exports.thenByDescending = orderby_1.thenByDescending;
const tsickle_forward_declare_55 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.orderby");
var pairwise_1 = goog.require('targets.es2015.cls.asynciterable.pairwise');
exports.pairwise = pairwise_1.pairwise;
const tsickle_forward_declare_56 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.pairwise");
var partition_1 = goog.require('targets.es2015.cls.asynciterable.partition');
exports.partition = partition_1.partition;
const tsickle_forward_declare_57 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.partition");
var pluck_1 = goog.require('targets.es2015.cls.asynciterable.pluck');
exports.pluck = pluck_1.pluck;
const tsickle_forward_declare_58 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.pluck");
var publish_1 = goog.require('targets.es2015.cls.asynciterable.publish');
exports.publish = publish_1.publish;
const tsickle_forward_declare_59 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.publish");
var race_1 = goog.require('targets.es2015.cls.asynciterable.race');
exports.race = race_1.race;
const tsickle_forward_declare_60 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.race");
var range_1 = goog.require('targets.es2015.cls.asynciterable.range');
exports.range = range_1.range;
const tsickle_forward_declare_61 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.range");
var reduce_1 = goog.require('targets.es2015.cls.asynciterable.reduce');
exports.reduce = reduce_1.reduce;
const tsickle_forward_declare_62 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.reduce");
var reduceright_1 = goog.require('targets.es2015.cls.asynciterable.reduceright');
exports.reduceRight = reduceright_1.reduceRight;
const tsickle_forward_declare_63 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.reduceright");
var repeat_1 = goog.require('targets.es2015.cls.asynciterable.repeat');
exports.repeat = repeat_1.repeat;
exports.repeatStatic = repeat_1.repeatStatic;
const tsickle_forward_declare_64 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.repeat");
var retry_1 = goog.require('targets.es2015.cls.asynciterable.retry');
exports.retry = retry_1.retry;
const tsickle_forward_declare_65 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.retry");
var reverse_1 = goog.require('targets.es2015.cls.asynciterable.reverse');
exports.reverse = reverse_1.reverse;
const tsickle_forward_declare_66 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.reverse");
var scan_1 = goog.require('targets.es2015.cls.asynciterable.scan');
exports.scan = scan_1.scan;
const tsickle_forward_declare_67 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.scan");
var scanright_1 = goog.require('targets.es2015.cls.asynciterable.scanright');
exports.scanRight = scanright_1.scanRight;
const tsickle_forward_declare_68 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.scanright");
var sequenceequal_1 = goog.require('targets.es2015.cls.asynciterable.sequenceequal');
exports.sequenceEqual = sequenceequal_1.sequenceEqual;
const tsickle_forward_declare_69 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.sequenceequal");
var share_1 = goog.require('targets.es2015.cls.asynciterable.share');
exports.share = share_1.share;
const tsickle_forward_declare_70 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.share");
var single_1 = goog.require('targets.es2015.cls.asynciterable.single');
exports.single = single_1.single;
const tsickle_forward_declare_71 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.single");
var skip_1 = goog.require('targets.es2015.cls.asynciterable.skip');
exports.skip = skip_1.skip;
const tsickle_forward_declare_72 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.skip");
var skiplast_1 = goog.require('targets.es2015.cls.asynciterable.skiplast');
exports.skipLast = skiplast_1.skipLast;
const tsickle_forward_declare_73 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.skiplast");
var skipuntil_1 = goog.require('targets.es2015.cls.asynciterable.skipuntil');
exports.skipUntil = skipuntil_1.skipUntil;
const tsickle_forward_declare_74 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.skipuntil");
var skipwhile_1 = goog.require('targets.es2015.cls.asynciterable.skipwhile');
exports.skipWhile = skipwhile_1.skipWhile;
const tsickle_forward_declare_75 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.skipwhile");
var slice_1 = goog.require('targets.es2015.cls.asynciterable.slice');
exports.slice = slice_1.slice;
const tsickle_forward_declare_76 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.slice");
var some_1 = goog.require('targets.es2015.cls.asynciterable.some');
exports.some = some_1.some;
const tsickle_forward_declare_77 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.some");
var startwith_1 = goog.require('targets.es2015.cls.asynciterable.startwith');
exports.startWith = startwith_1.startWith;
const tsickle_forward_declare_78 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.startwith");
var sum_1 = goog.require('targets.es2015.cls.asynciterable.sum');
exports.sum = sum_1.sum;
const tsickle_forward_declare_79 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.sum");
var take_1 = goog.require('targets.es2015.cls.asynciterable.take');
exports.take = take_1.take;
const tsickle_forward_declare_80 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.take");
var takelast_1 = goog.require('targets.es2015.cls.asynciterable.takelast');
exports.takeLast = takelast_1.takeLast;
const tsickle_forward_declare_81 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.takelast");
var takeuntil_1 = goog.require('targets.es2015.cls.asynciterable.takeuntil');
exports.takeUntil = takeuntil_1.takeUntil;
const tsickle_forward_declare_82 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.takeuntil");
var takewhile_1 = goog.require('targets.es2015.cls.asynciterable.takewhile');
exports.takeWhile = takewhile_1.takeWhile;
const tsickle_forward_declare_83 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.takewhile");
var tap_1 = goog.require('targets.es2015.cls.asynciterable.tap');
exports.tap = tap_1.tap;
const tsickle_forward_declare_84 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.tap");
var throttle_1 = goog.require('targets.es2015.cls.asynciterable.throttle');
exports.throttle = throttle_1.throttle;
const tsickle_forward_declare_85 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.throttle");
var throw_1 = goog.require('targets.es2015.cls.asynciterable.throw');
exports._throw = throw_1._throw;
const tsickle_forward_declare_86 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.throw");
var toarray_1 = goog.require('targets.es2015.cls.asynciterable.toarray');
exports.toArray = toarray_1.toArray;
const tsickle_forward_declare_87 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.toarray");
var tomap_1 = goog.require('targets.es2015.cls.asynciterable.tomap');
exports.toMap = tomap_1.toMap;
const tsickle_forward_declare_88 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.tomap");
var toobservable_1 = goog.require('targets.es2015.cls.asynciterable.toobservable');
exports.toObservable = toobservable_1.toObservable;
const tsickle_forward_declare_89 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.toobservable");
var toset_1 = goog.require('targets.es2015.cls.asynciterable.toset');
exports.toSet = toset_1.toSet;
const tsickle_forward_declare_90 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.toset");
var union_1 = goog.require('targets.es2015.cls.asynciterable.union');
exports.union = union_1.union;
const tsickle_forward_declare_91 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.union");
var while_1 = goog.require('targets.es2015.cls.asynciterable.while');
exports._while = while_1._while;
const tsickle_forward_declare_92 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.while");
var zip_1 = goog.require('targets.es2015.cls.asynciterable.zip');
exports.zip = zip_1.zip;
const tsickle_forward_declare_93 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable.zip");
