// ix/targets/es2015/cls/asynciterable/concat.js
goog.module('targets.es2015.cls.asynciterable.concat'); exports = {}; var module = {id: 'targets/es2015/cls/asynciterable/concat.js'};/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

var asynciterable_1 = goog.require('targets.es2015.cls.asynciterable');
const tsickle_forward_declare_1 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.asynciterable");
/**
 * @template TSource
 */
class ConcatAllAsyncIterable extends asynciterable_1.AsyncIterableX {
    /**
     * @param {!AsyncIterable<!AsyncIterable<TSource>>} source
     */
    constructor(source) {
        super();
        this._source = source;
    }
    /**
     * @return {!AsyncIterableIterator<TSource>}
     */
    [Symbol.asyncIterator]() {
        return __asyncGenerator(this, arguments, function* _a() {
            try {
                for (var _a = __asyncValues(this._source), _b; _b = yield __await(_a.next()), !_b.done;) {
                    let /** @type {!AsyncIterable<TSource>} */ outer = yield __await(_b.value);
                    try {
                        for (var outer_1 = __asyncValues(outer), outer_1_1; outer_1_1 = yield __await(outer_1.next()), !outer_1_1.done;) {
                            let /** @type {TSource} */ item = yield __await(outer_1_1.value);
                            yield item;
                        }
                    }
                    catch (e_1_1) { e_1 = { error: e_1_1 }; }
                    finally {
                        try {
                            if (outer_1_1 && !outer_1_1.done && (_c = outer_1.return)) yield __await(_c.call(outer_1));
                        }
                        finally { if (e_1) throw e_1.error; }
                    }
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (_b && !_b.done && (_d = _a.return)) yield __await(_d.call(_a));
                }
                finally { if (e_2) throw e_2.error; }
            }
            var e_2, _d, e_1, _c;
        });
    }
}
function ConcatAllAsyncIterable_tsickle_Closure_declarations() {
    /** @type {!AsyncIterable<!AsyncIterable<TSource>>} */
    ConcatAllAsyncIterable.prototype._source;
}
/**
 * @template TSource
 */
class ConcatAsyncIterable extends asynciterable_1.AsyncIterableX {
    /**
     * @param {!Iterable<!AsyncIterable<TSource>>} source
     */
    constructor(source) {
        super();
        this._source = source;
    }
    /**
     * @return {!AsyncIterableIterator<TSource>}
     */
    [Symbol.asyncIterator]() {
        return __asyncGenerator(this, arguments, function* _a() {
            for (let /** @type {!AsyncIterable<TSource>} */ outer of this._source) {
                try {
                    for (var outer_2 = __asyncValues(outer), outer_2_1; outer_2_1 = yield __await(outer_2.next()), !outer_2_1.done;) {
                        let /** @type {TSource} */ item = yield __await(outer_2_1.value);
                        yield item;
                    }
                }
                catch (e_3_1) { e_3 = { error: e_3_1 }; }
                finally {
                    try {
                        if (outer_2_1 && !outer_2_1.done && (_a = outer_2.return)) yield __await(_a.call(outer_2));
                    }
                    finally { if (e_3) throw e_3.error; }
                }
            }
            var e_3, _a;
        });
    }
}
function ConcatAsyncIterable_tsickle_Closure_declarations() {
    /** @type {!Iterable<!AsyncIterable<TSource>>} */
    ConcatAsyncIterable.prototype._source;
}
/**
 * @template TSource
 * @param {!AsyncIterable<!AsyncIterable<TSource>>} source
 * @return {!tsickle_forward_declare_1.AsyncIterableX<TSource>}
 */
function concatAll(source) {
    return new ConcatAllAsyncIterable(source);
}
exports.concatAll = concatAll;
/**
 * @template TSource
 * @param {!Iterable<!AsyncIterable<TSource>>} source
 * @return {!tsickle_forward_declare_1.AsyncIterableX<TSource>}
 */
function _concatAll(source) {
    return new ConcatAsyncIterable(source);
}
exports._concatAll = _concatAll;
/**
 * @template T
 * @param {!AsyncIterable<T>} source
 * @param {...!AsyncIterable<T>} args
 * @return {!tsickle_forward_declare_1.AsyncIterableX<T>}
 */
function concat(source, ...args) {
    return new ConcatAsyncIterable([source, ...args]);
}
exports.concat = concat;
/**
 * @template T
 * @param {...!AsyncIterable<T>} args
 * @return {!tsickle_forward_declare_1.AsyncIterableX<T>}
 */
function concatStatic(...args) {
    return new ConcatAsyncIterable(args);
}
exports.concatStatic = concatStatic;
