// ix/targets/es2015/cls/asynciterable/toobservable.js
goog.module('targets.es2015.cls.asynciterable.toobservable'); exports = {}; var module = {id: 'targets/es2015/cls/asynciterable/toobservable.js'};/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

const tsickle_forward_declare_1 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.observer");
goog.require("_Users.ptaylor.dev.ixjs.src.observer"); // force type-only module to be loaded
const tsickle_forward_declare_2 = goog.forwardDeclare("_Users.ptaylor.dev.ixjs.src.subscription");
goog.require("_Users.ptaylor.dev.ixjs.src.subscription"); // force type-only module to be loaded
/**
 * @implements {tsickle_forward_declare_2.Subscription}
 */
class BooleanSubscription {
    constructor() {
        this.isUnsubscribed = false;
    }
    /**
     * @return {void}
     */
    unsubscribe() {
        this.isUnsubscribed = true;
    }
}
function BooleanSubscription_tsickle_Closure_declarations() {
    /** @type {boolean} */
    BooleanSubscription.prototype.isUnsubscribed;
}
/**
 * @template TSource
 * @implements {tsickle_forward_declare_1.Observable}
 */
class AsyncIterableObservable {
    /**
     * @param {!AsyncIterable<TSource>} source
     */
    constructor(source) {
        this._source = source;
    }
    /**
     * @param {!tsickle_forward_declare_1.Observer<TSource>} observer
     * @return {!BooleanSubscription}
     */
    subscribe(observer) {
        const /** @type {!BooleanSubscription} */ subscription = new BooleanSubscription();
        const /** @type {!AsyncIterator<TSource>} */ it = this._source[Symbol.asyncIterator]();
        const /** @type {function(): void} */ f = () => {
            it.next()
                .then(({ value, done }) => {
                if (!subscription.isUnsubscribed) {
                    if (done) {
                        observer.complete();
                    }
                    else {
                        observer.next(value);
                        f();
                    }
                }
            })
                .catch(err => {
                if (!subscription.isUnsubscribed) {
                    observer.error(err);
                }
            });
        };
        f();
        return subscription;
    }
}
function AsyncIterableObservable_tsickle_Closure_declarations() {
    /** @type {!AsyncIterable<TSource>} */
    AsyncIterableObservable.prototype._source;
}
/**
 * @template TSource
 * @param {!AsyncIterable<TSource>} source
 * @return {!tsickle_forward_declare_1.Observable<TSource>}
 */
function toObservable(source) {
    return new AsyncIterableObservable(source);
}
exports.toObservable = toObservable;
